shopfloor_packing_info/hooks.py

```diff
@@ -0,0 +1,131 @@
+#        of the column that is going to be removed when the update
+#        on 'shopfloor' is done. Since updating & installing at the
+#        same time sometimes yields to problems in databases having
+#        already installed the module that has the field to be removed
+#        defined in some views, it deletes it from the views defined
+#        in the original module.
+#     2. post-init will get the values from that temporary column,
+#        (because the update would have removed the original column)
+#        and use them to populate the new records for the new model,
+#        and link them to the corresponding customers. The temporal
+#        column will be removed afterwards.
+
+
+import hashlib
+import os
+
+from odoo.api import SUPERUSER_ID, Environment
+
+
+def pre_init_hook(cr):
+    if _column_exists(cr, "res_partner", "shopfloor_packing_info"):
+        cr.execute(
+            """
+            ALTER TABLE res_partner
+            ADD COLUMN shopfloor_packing_info_tmp TEXT;
+
+            UPDATE res_partner
+            SET shopfloor_packing_info_tmp=shopfloor_packing_info;"""
+        )
+        cr.execute(
+            """
+            DELETE FROM ir_ui_view WHERE id IN (
+              SELECT v.id
+              FROM ir_ui_view v, ir_model_data m
+              WHERE v.arch_db ILIKE '%shopfloor_packing_info%'
+                AND m.res_id = v.id
+                AND m.module = 'shopfloor'
+                AND m.model = 'ir.ui.view');
+            """
+        )
+
+
+def post_init_hook(cr, registry):
+    if _column_exists(cr, "res_partner", "shopfloor_packing_info_tmp"):
+        copy_into_new_model(cr)
+        cr.execute(
+            """
+            ALTER TABLE res_partner
+            DROP COLUMN shopfloor_packing_info_tmp;"""
+        )
+
+
+def _column_exists(cr, table_name, column_name):
+    # See https://stackoverflow.com/a/9991073
+    cr.execute(
+        """
+        SELECT count(attname)
+        FROM pg_attribute
+        WHERE attrelid = ( SELECT oid FROM pg_class WHERE relname = %s )
+          AND attname = %s;""",
+        (table_name, column_name),
+    )
+    return cr.fetchone()[0] == 1
+
+
+def copy_into_new_model(cr):
+    env = Environment(cr, SUPERUSER_ID, {})
+    shopfloor_packing_info = env["shopfloor.packing.info"]
+    res_partner = env["res.partner"]
+    ir_model_data = env["ir.model.data"]
+
+    env.cr.execute(
+        """
+        SELECT id, COALESCE(shopfloor_packing_info_tmp, '') AS packing_info_old
+        FROM res_partner
+        WHERE active = TRUE;
+        """
+    )
+
+    for row in env.cr.dictfetchall():
+        partner_id = int(row["id"])
+        packing_info_full = row["packing_info_old"].strip()
+        if packing_info_full:
+            partner = res_partner.browse(partner_id)
+
+            # We assume the name is the first line, if there are several.
+            linesep_pos = packing_info_full.find(os.linesep)
```

Comment:
just `packing_info_full.splitlines()`?

Reply:
In this case it doesn't matter because it's a code that is executed once everytime, but I didn't use splitlines nor split because it's much slower. See:

```python
>>> import timeit
>>>
>>> timeit.timeit("x.find('\\n')", setup="x = 'hi\\nhow\\nare\\nyou\\ndoing'")
0.10260158000164665
>>> timeit.timeit("x.find('\\n')", setup="x = 'hi\\nhow\\nare\\nyou\\ndoing'")
0.09811641899796086
>>> timeit.timeit("x.find('\\n')", setup="x = 'hi\\nhow\\nare\\nyou\\ndoing'")
0.09946276299888268
>>> 
>>> timeit.timeit('x.splitlines()', setup="x = 'hi\\nhow\\nare\\nyou\\ndoing'")
0.23253642599956947
>>> timeit.timeit('x.splitlines()', setup="x = 'hi\\nhow\\nare\\nyou\\ndoing'")
0.23402035599792725
>>> timeit.timeit('x.splitlines()', setup="x = 'hi\\nhow\\nare\\nyou\\ndoing'")
0.236937740999565
```